src/bin/jcli_app/utils/io.rs

```diff
@@ -1,37 +1,48 @@
+use std::fs::File;
+use std::io::{stdin, stdout, BufRead, BufReader, Error as IoError, Write};
 use std::path::Path;
 
-/// open the given file path as a writable stream, or stdout if no path
+/// open the given file path as a readable stream, or stdin if no path
 /// provided
-pub fn open_file_write<P: AsRef<Path>>(path: &Option<P>) -> Box<dyn std::io::Write> {
-    if let Some(path) = path {
-        Box::new(
-            std::fs::OpenOptions::new()
-                .create(true)
-                .write(true)
-                .read(false)
-                .append(false)
-                .open(path)
-                .unwrap(),
-        )
-    } else {
-        Box::new(std::io::stdout())
+pub fn open_file_read<P: AsRef<Path>>(path: &Option<P>) -> Result<impl BufRead, Error> {
+    match path {
+        Some(path) => {
+            let reader = File::open(path).map_err(|e| Error::new_open_file_read(path, e))?;
+            Ok(Box::new(BufReader::new(reader)) as Box<dyn BufRead>)
+        }
+        None => Ok(Box::new(BufReader::new(stdin())) as Box<dyn BufRead>),
     }
 }
 
-/// open the given file path as a readable stream, or stdin if no path
+/// open the given file path as a writable stream, or stdout if no path
 /// provided
-pub fn open_file_read<P: AsRef<Path>>(path: &Option<P>) -> Box<dyn std::io::BufRead> {
-    if let Some(path) = path {
-        Box::new(std::io::BufReader::new(
-            std::fs::OpenOptions::new()
-                .create(false)
-                .write(false)
-                .read(true)
-                .append(false)
-                .open(path)
-                .unwrap(),
-        ))
-    } else {
-        Box::new(std::io::BufReader::new(std::io::stdin()))
+pub fn open_file_write<P: AsRef<Path>>(path: &Option<P>) -> Result<impl Write, Error> {
+    match path {
+        Some(path) => {
+            let writer = File::create(path).map_err(|e| Error::new_open_file_write(path, e))?;
```

Comment:
I prefer the `OpenOptions` method. we see explicitly the options when opening the file. What is the rational for the change of function ?

Reply:
There was a subtle bug in open_file_write: the truncate flag was not set. If we try to overwrite a file containing 20 bytes with 10 bytes of JCLI output, the file will still be 20 bytes, 10 new and 10 old. This bug would not be possible if we used the methods provided by std lib, which do what we want, but are guaranteed to be correct. Also the old version was unnecessarily noisy and verbose.

Reply:
1. indeed the `truncate` option was not set, this was the bug, it should have been the only line changing here.
2. verbose is fine, especially if it allows to be precise, this was the case here. It was precise and it was easy to identify the error.
3. OpenOptions is the method provided by rust `std`.

Please revert change and fix the missing truncate and the error report as 2 different commit:

1. for the missing truncate;
2. one to add the Error